lib/pbench/test/unit/agent/test_utils.py

```diff
@@ -253,3 +260,286 @@ def test_methods(monkeypatch):
+                "nis": "hostname[mock]: Local domain name not set",
+                "short": "agent",
+            },
+            "hostname": "remote-a.example.com",
+            "installs": {"mpstat": [0, "mpstat: pbench-sysstat-12.0.3-1 is installed"]},
+            "kind": "tm",
+            "label": "",
+            "noop_tools": [],
+            "persistent_tools": [],
+            "pid": "NNNNN",
+            "posted": "dormant",
+            "seqno": "",
+            "sha1": "(unknown)",
+            "tools": {"mpstat": "--interval=42 --options=forty-two"},
+            "transient_tools": [],
+            "version": "(unknown)",
+        },
+        "remote-b.example.com": {
+            "failed_tools": ["mpstat"],
+            "hostdata": {
+                "alias": "bond",
+                "all-fqdns": "agent.example.com bond.example.com",
+                "all-ip-addresses": "192.168.0.1 172.0.0.1 127.0.0.1",
+                "domain": "example.com",
+                "fqdn": "agent.example.com",
+                "ip-address": "172.0.0.1 127.0.0.1",
+                "nis": "hostname[mock]: Local domain name not set",
+                "short": "agent",
+            },
+            "hostname": "remote-b.example.com",
+            "installs": {
+                "mpstat": [0, "mpstat: pbench-sysstat-12.0.3-1 is installed"],
+                "node-exporter": [0, "node_exporter tool properly installed"],
+            },
+            "kind": "tm",
+            "label": "blue",
+            "noop_tools": [],
+            "persistent_tools": ["node-exporter"],
+            "pid": "NNNNN",
+            "posted": "dormant",
+            "seqno": "",
+            "sha1": "(unknown)",
+            "tools": {
+                "mpstat": "--interval=42 --options=forty-two",
+                "node-exporter": "--inst=/var/tmp/pbench-test-utils/opt/pbench-agent/unittest-scripts",
+            },
+            "transient_tools": [],
+            "version": "(unknown)",
+        },
+    }
+
+    def test_check_failed_tool_init(self):
+        status = check_failed_tool_init(self.mock_tms_failed_data)
+        assert status == "failed"
+
+    def test_check_success_tool_init(self):
+        status = check_failed_tool_init(self.mock_tms_success_data)
+        assert status == "success"
+
+    def test_log_failed_tool_init_multiple_hosts(self):
+        logger = logging.getLogger("test_log_methods")
+        failure_count = log_failed_tool_init(logger, self.mock_tms_data_multiple_hosts)
+        assert failure_count == 2
+
+    def test_log_failed_tool_init_single_host(self):
+        logger = logging.getLogger("test_log_methods")
+        failure_count = log_failed_tool_init(logger, self.mock_tms_data_single_host)
+        assert failure_count == 1
+
+    def test_log_success_tool_init(self):
+        logger = logging.getLogger("test_log_methods")
+        failure_count = log_failed_tool_init(logger, self.mock_tms_log_data)
+        assert failure_count == 0
```

Comment:
Arguably, we should test `log_failed_tool_init()` with the same ten scenarios that I proposed for `check_failed_tool_init()`.  However, for each remote where there is a tool failure, we really should test an analogous ten sub-scenarios (i.e., one, two, or three tools; one, two, or three failures among them; with the failures occurring in each possible position).

I think there is a nice, programmatic way to generate the cases, so we could use `@pytest.mark.parametrize()` with an argument which is a generator function instead of a constant list.  But, I think I'll leave coding that up to tomorrow. 😴

Reply:
If I move module `log_failed_tool_init` back to `pbench-tool-meister-start` then, I believe I won't be able to add pytests for this module.

Reply:
Yes, that's true, and it's a valid point. (I'm going to have some nasty rebases, I expect, but I'm working on refactoring `pbench-tool-meister-{start|stop}` into click-like trampoline files with all the logic in testable modules; and this will likely go into a shared based class rather than this generic utilities module.)

I think that if you take my suggestions about merging these twin functions, this will be less of an issue since the utility is used by both TDS and start. 😃

Reply:
It needs to be removed from TDS entirely.

There is no need for TDS to use either of these functions, they are entirely for the start mechanism to use because they are about messaging the user.

And the plain logging is not what we want, we want to use `error_log`.  The logging methods only dump to the pbench.log file, I believe.

Reply:
I'll post a PR with my thinking.

Reply:
I don't think that's quite true; the "check" variety is used by TDS in processing the TM confirmations to determine whether the message back to the CLI is "success" or "failure".

Now I suppose this could be simplified, because the individual TM clients should be reporting "failure" if there are any problems, which really means TDS only needs to report failure if at least one of the clients did; it arguably shouldn't have any need to re-analyze the data.

Reply:
I know that is this code is proposing to do, but it is not necessary.  I'll be posting a PR shortly.